### ISSUE TITLE
SC2154 "referenced but not assigned" for variables assigned and referenced in a trap

### ISSUE THREAD

**Issue by Ben-Voris** [koalaman/shellcheck#2533]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2154, SC2154
- My shellcheck version (`shellcheck --version` or 'online'): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/env bash
# Fail on errors, use of an undefined variable, and if any part of pipe fails.
set -euo pipefail
# Pass error handling to subshells, like in $()
set -E
# Trap errors
trap '{ ec=$? ;
        ln=$LINENO ;
        echo -e "\n\n ** Exit code ${ec} occured on line ${ln} **\n\n" ;
        exit ${ec}; }' ERR

PROG=test
echo Building ${PROG} version "$(git describe --dirty --always --tags)"

exit 0
```

#### Here's what shellcheck currently says:
```
Line 7	SC2154: ec is referenced but not assigned.
Line 7	SC2154: ln is referenced but not assigned (for output from commands, use "$(ln ...)" ).
```
#### Here's what I wanted or expected to see:

TODO: Describe expected/desired output

To me, it appears that both `ec` and `ln` are assigned values.

**Comment by Ben-Voris** [comment]
I now see it's a duplicate of https://github.com/koalaman/shellcheck/issues/1299 and https://github.com/koalaman/shellcheck/issues/839
